File: 백준/Gold/1027. 고층 건물/고층 건물.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N, ans = 0;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        N = Integer.parseInt(br.readLine());
        arr = new int[N + 1]; // 배열 초기화 필수
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; ++i) arr[i] = Integer.parseInt(st.nextToken());
        for (int i = 1; i <= N; ++i) {
            int cnt = 0;
            for (int j = 1; j <= N; ++j) {
                if (i == j) continue;
                if (findBuilding(i, j)) cnt++;
            }
            ans = Math.max(cnt, ans);
        }
        System.out.println(ans);
    }

    public static boolean findBuilding(int i, int j) {
        if (i > j) {
            int temp = i;
            i = j;
            j = temp;
        }
        for (int k = i + 1; k < j; ++k) {
            double height = arr[i] + (double)(arr[i] - arr[j]) * (k - i) / (i - j);
            if (height <= arr[k]) return false;
        }
        return true;
    }
}

File: 백준/Gold/1027. 고층 건물/고층 건물.cc
// 풀이 시간: 1h30m55s54
// 시간 복잡도: O(N^2)
// 공간 복잡도: O(N)

#include <iostream>
#include <cmath>
#include <algorithm>
using namespace std;
int N, ans = 0;
int arr[51] = { 0, };

bool findBuilding(int i, int j) {
    if (i > j) swap(i, j); // 항상 (i < j)로 계산
    for (int k = i + 1; k < j; ++k) { 
        double height = arr[i] + double(arr[i] - arr[j]) * (k - i) / (i - j); // 높이 (타입 유의)
        if (height <= arr[k]) return false; // 해당 선분에 어떤 고층빌딩도 지나가면 안되기 때문에 같은 높이도 안됨
    }
    return true;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0); cout.tie(0);
    cin >> N;
    for (int i = 1; i <= N; ++i) cin >> arr[i];
    for (int i = 1; i <= N; ++i) {
        int cnt = 0;
        for (int j = 1; j <= N; ++j) {
            if (i == j) continue;
            if (findBuilding(i, j)) cnt++; // 해당 빌딩이 보이면 카운트
        }
        ans = max(cnt, ans);
    }
    cout << ans;
    return 0;
}